File: 153-find-minimum-in-rotated-sorted-array/solution.py
class Solution:
    def findMin(self, nums: list[int]) -> int:
        left, right = 0, len(nums) - 1

        # Standard Binary Search setup
        while left < right:
            mid = left + (right - left) // 2

            # If the middle element is greater than the rightmost element,
            # it implies the rotation point (and thus the minimum) is to the right.
            if nums[mid] > nums[right]:
                left = mid + 1
            # If the middle element is smaller than the rightmost element,
            # the minimum is either at mid or to the left of mid.
            else:
                right = mid
        
        # After the loop, left == right, pointing to the minimum element.
        return nums[left]

File: 153-find-minimum-in-rotated-sorted-array/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0;
        int right = nums.size() - 1;

        // Binary search to find the minimum element
        while (left < right) {
            int mid = left + (right - left) / 2;

            // If mid element is greater than the rightmost element, 
            // the minimum must be in the right half.
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } 
            // Otherwise, the minimum is in the left half (including mid).
            else {
                right = mid;
            }
        }

        // When left == right, we have found the minimum element.
        return nums[left];
    }
};